### ISSUE TITLE
Text processing bugs

### ISSUE THREAD

**Issue by Gillibald** [AvaloniaUI/Avalonia#7502]
This issue helps track bugs related to recent text processing changes. Feel free to post your findings here.

**Comment by Gillibald** [comment]
Posted by @danielmayost

There are a few more, if editing text is an ongoing work, it can be addressed after merge.

(from ControlCatalog):
1. `<TextBox Width="200" Text="RightToLeft text" FlowDirection="RightToLeft" />` has bouncy selection:
<img src="https://user-images.githubusercontent.com/41772276/151871441-ece7112b-1163-423f-b932-b2269adc7b3c.gif" width=200 height=30>

2. a `אבג 1` not selected:
<img src="https://user-images.githubusercontent.com/41772276/151871777-a5a623b8-dc68-4fbf-bee9-98a165060ac0.gif" width=200 height=30>

3. `<TextBox Width="200" Text="אבג 1" TextAlignment="Left" />` The cursor goes backwards by typing a space:
<img src="https://user-images.githubusercontent.com/41772276/151872191-60ec6df7-9b9d-4db2-833a-94d39107aba7.gif" width=200 height=30>

4. `<TextBox Width="200" FlowDirection="RightToLeft" />` Type `א` and then `a` changes the baseline:
<img src="https://user-images.githubusercontent.com/41772276/151872935-97b8a07a-a606-40f7-a3d0-5248641b2e1a.gif" width=200 height=30>

_Originally posted by @danielmayost in https://github.com/AvaloniaUI/Avalonia/issues/4893#issuecomment-1026208611_

**Comment by Gillibald** [comment]
2. is fixed

**Comment by danielmayost** [comment]
5. DataGrid Editable has no cursor:
<img src="https://user-images.githubusercontent.com/41772276/152021901-0cb0f249-81d0-44a5-8623-95d54884c14c.gif" width=300 height=150>

6. when input `אAאAאA` selected all, there is a break between the glyphs:
<img src="https://user-images.githubusercontent.com/41772276/152022506-d15ef70f-26fd-47a2-a3db-7901f2bd65e3.png" width=200 height=50>

7. selection using the keyboard when `FlowDirection="RightToLeft"`, should be reversed like notepad (left arrow advances in logical order).

**Comment by Gillibald** [comment]
6. Selection rectangles are currently not merged will change that

**Comment by danielmayost** [comment]
yes, should be like windows notepad etc.

**Comment by maxkatz6** [comment]
TextBox with AcceptsReturn="True" doesn't receive any input after couple of empty lines:
https://user-images.githubusercontent.com/3163374/152086985-32b8c842-25e7-4eff-970b-6321427f44e0.mp4

**Comment by hez2010** [comment]
[ ] Input position doesn't respect caret movement (up/down) while using keyboard: I moved the caret to upper lines using keyboard, but when I started to input, the caret jumped to the original position
[ ] Text selection is broken

https://user-images.githubusercontent.com/14960345/152672625-31e4b41b-0c7c-4f35-9130-c8de39383582.mp4

**Comment by danielmayost** [comment]
- [ ] not selected yet (`ABC אבג`):
![test](https://user-images.githubusercontent.com/41772276/152942900-d1d04f9f-f748-4627-9d5f-1e0e4c26c9ad.gif)

**Comment by adirh3** [comment]
Selection with Ctrl + Shift + Left/Right arrow keys is not working; it just moves the caret.

**Comment by luthfiampas** [comment]
Selection with shift + left click behavior is broken because caret index was being changed.

![record-(2)](https://user-images.githubusercontent.com/57059775/153443057-1fc9731c-57da-49b3-9763-e769d14904c4.gif)

https://github.com/AvaloniaUI/Avalonia/pull/7280/files#diff-1b561a0dd4b8160e998d7346ff6b4d5c75ceb71edc990b1b302468cbabfe7a65R1012

**Comment by adirh3** [comment]
When setting RTL flow direction causing controls to stretch indefinitely.
Before - 
![image](https://user-images.githubusercontent.com/27368554/153657932-c12466a6-5a9d-4b8c-aca3-b1a42df83be0.png)

After - 
![image](https://user-images.githubusercontent.com/27368554/153658157-a5f34101-e4f3-429e-8578-b71d58e7a696.png)

**Comment by Gillibald** [comment]
> When setting RTL flow direction causing controls to stretch indefinitely. Before - ![image](https://user-images.githubusercontent.com/27368554/153657932-c12466a6-5a9d-4b8c-aca3-b1a42df83be0.png)
> 
> After - ![image](https://user-images.githubusercontent.com/27368554/153658157-a5f34101-e4f3-429e-8578-b71d58e7a696.png)

Must be something special about Menu/Flyout will try to reproduce it

**Comment by adirh3** [comment]
> > When setting RTL flow direction causing controls to stretch indefinitely. Before - ![image](https://user-images.githubusercontent.com/27368554/153657932-c12466a6-5a9d-4b8c-aca3-b1a42df83be0.png)
> > After - ![image](https://user-images.githubusercontent.com/27368554/153658157-a5f34101-e4f3-429e-8578-b71d58e7a696.png)
> 
> Must be something special about Menu/Flyout will try to reproduce it

It seems to happen to almost everything. I might be using it wrong? Tried setting `FlowDirection` to all `TextBlock`s using a `Style`

Before:
![image](https://user-images.githubusercontent.com/27368554/153671827-c732015d-1899-4f74-ae10-81b9744a3003.png)

After: 
<img width="587" alt="image" src="https://user-images.githubusercontent.com/27368554/153672078-404d18a3-c2ff-45f2-bfd8-785f0958e32d.png">

**Comment by adirh3** [comment]
@Gillibald sharing here a few exceptions from users, I haven't personally reproduced all of them, but they do not happen on previous version of my app (only change is Avalonia from 0.10.12 -> master).

```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'length')

Avalonia.Media.TextFormatting
FormattedTextSource.CreateTextStyleRun (ReadOnlySlice`1 text, TextRunProperties defaultProperties, IReadOnlyList`1 textModifier)
Avalonia.Media.TextFormatting
FormattedTextSource.GetTextRun (Int32 textSourceIndex)
Avalonia.Media.TextFormatting
TextFormatterImpl.FetchTextRuns (ITextSource textSource, Int32 firstTextSourceIndex, TextEndOfLine& endOfLine, TextRange& textRange)
Avalonia.Media.TextFormatting
TextFormatterImpl.FormatLine (ITextSource textSource, Int32 firstTextSourceIndex, Double paragraphWidth, TextParagraphProperties paragraphProperties, TextLineBreak previousLineBreak)
Avalonia.Media.TextFormatting
TextLayout.CreateTextLines ()
Avalonia.Controls.Presenters
TextPresenter.CreateTextLayoutInternal (Size constraint, String text, Typeface typeface, IReadOnlyList`1 textStyleOverrides)
Avalonia.Controls.Presenters
TextPresenter.CreateTextLayout ()
Avalonia.Controls.Presenters
TextPresenter.get_TextLayout ()
Avalonia.Controls.Presenters
TextPresenter.MoveCaretToTextPosition (Int32 textPosition, Boolean trailingEdge)
Avalonia
DirectProperty`2.InvokeSetter (IAvaloniaObject instance, BindingValue`1 value)
Avalonia
AvaloniaObject.SetDirectValueUnchecked[T] (DirectPropertyBase`1 property, BindingValue`1 value)
Avalonia.AvaloniaObject
DirectBindingSubscription`1.OnNext (BindingValue`1 value)
Avalonia.Reactive
TypedBindingAdapter`1.OnNext (BindingValue`1 value)
Avalonia.Reactive
BindingValueAdapter`1.OnNext (T value)
Avalonia
AvaloniaObject.RaisePropertyChanged[T] (AvaloniaPropertyChangedEventArgs`1 change)
Avalonia.Controls
TextBox.set_CaretIndex (Int32 value)
Avalonia.Controls
TextBox.DeleteSelection (Boolean raiseTextChanged)
Avalonia.Controls
TextBox.OnKeyDown (KeyEventArgs e)
System.Reactive.Subjects
Subject`1.OnNext (T value) /_/Rx.NET/Source/src/System.Reactive/Subjects/Subject.cs:145
Avalonia.Interactivity
EventRoute.RaiseEventImpl (RoutedEventArgs e)
Avalonia.Interactivity
EventRoute.RaiseEvent (IInteractive source, RoutedEventArgs e)
Avalonia.Interactivity
Interactive.RaiseEvent (RoutedEventArgs e)
Avalonia.Input
KeyboardDevice.ProcessRawEvent (RawInputEventArgs e)
Avalonia.Input
InputManager.ProcessInput (RawInputEventArgs e)
Avalonia.Win32
WindowImpl.AppWndProc (IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
Avalonia.Win32
WindowImpl.WndProc (IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
Avalonia.Win32.Interop
UnmanagedMethods.DispatchMessage (MSG& lpmsg)
Avalonia.Win32
Win32Platform.RunLoop (CancellationToken cancellationToken)
Avalonia.Threading
Dispatcher.MainLoop (CancellationToken cancellationToken)
Avalonia.Controls.ApplicationLifetimes
ClassicDesktopStyleApplicationLifetime.Start (String[] args)
```

```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'length')

Avalonia.Utilities
ReadOnlySlice`1.Take (Int32 length)
Avalonia.Media.TextFormatting
FormattedTextSource.CreateTextStyleRun (ReadOnlySlice`1 text, TextRunProperties defaultProperties, IReadOnlyList`1 textModifier)
Avalonia.Media.TextFormatting
TextFormatterImpl.FetchTextRuns (ITextSource textSource, Int32 firstTextSourceIndex, TextEndOfLine& endOfLine, TextRange& textRange)
Avalonia.Media.TextFormatting
TextLayout.CreateTextLines ()
Avalonia.Media.TextFormatting.TextLayout
Avalonia.Media.TextFormatting.TextLayout..ctor(String text, Typeface typeface, Double fontSize, IBrush foreground, TextAlignment textAlignment, TextWrapping textWrapping, TextTrimming textTrimming, TextDecorationCollection textDecorations, FlowDirection flowDirection, Double maxWidth, Double maxHeight, Double lineHeight, Int32 maxLines, IReadOnlyList`1 textStyleOverrides)
Avalonia.Controls.Presenters
TextPresenter.CreateTextLayoutInternal (Size constraint, String text, Typeface typeface, IReadOnlyList`1 textStyleOverrides)
Avalonia.Controls.Presenters
TextPresenter.CreateTextLayout ()
Avalonia.Controls.Presenters
TextPresenter.get_TextLayout ()
Avalonia.Controls.Presenters
TextPresenter.MoveCaretToTextPosition (Int32 textPosition, Boolean trailingEdge)
Avalonia.Controls.Presenters
TextPresenter.set_CaretIndex (Int32 value)
Avalonia.Controls.Presenters.TextPresenter.<>c
<.cctor>b__24_1 (TextPresenter o, Int32 v)
Avalonia
DirectProperty`2.InvokeSetter (IAvaloniaObject instance, BindingValue`1 value)
Avalonia
AvaloniaObject.SetDirectValueUnchecked[T] (DirectPropertyBase`1 property, BindingValue`1 value)
Avalonia.AvaloniaObject
DirectBindingSubscription`1.OnNext (BindingValue`1 value)
Avalonia.Reactive
SingleSubscriberObservableBase`1.PublishNext (T value)
Avalonia.Reactive
TypedBindingAdapter`1.OnNext (BindingValue`1 value)
Avalonia.Reactive
BindingValueAdapter`1.OnNext (T value)
Avalonia.Data
TemplateBinding.TemplatedParentPropertyChanged (Object sender, AvaloniaPropertyChangedEventArgs e)
Avalonia
AvaloniaObject.RaisePropertyChanged[T] (AvaloniaPropertyChangedEventArgs`1 change)
Avalonia
AvaloniaObject.SetAndRaise[T] (AvaloniaProperty`1 property, T& field, T value)
Avalonia.Controls
TextBox.set_CaretIndex (Int32 value)
Avalonia.Controls
TextBox.DeleteSelection (Boolean raiseTextChanged)
Avalonia.Controls
TextBox.OnKeyDown (KeyEventArgs e)
Avalonia.Input.InputElement.<>c
<.cctor>b__32_2 (InputElement x, KeyEventArgs e)
System.Reactive
ObserverBase`1.OnNext (T value) /_/Rx.NET/Source/src/System.Reactive/ObserverBase.cs:32
Avalonia.Interactivity
EventRoute.RaiseEventImpl (RoutedEventArgs e)
Avalonia.Interactivity
Interactive.RaiseEvent (RoutedEventArgs e)
Avalonia.Input
KeyboardDevice.ProcessRawEvent (RawInputEventArgs e)
Avalonia.Controls
TopLevel.HandleInput (RawInputEventArgs e)
Avalonia.Win32
WindowImpl.AppWndProc (IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
Avalonia.Win32
WindowImpl.WndProc (IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
Avalonia.Win32.Interop
UnmanagedMethods.DispatchMessage (MSG& lpmsg)
Avalonia.Win32
Win32Platform.RunLoop (CancellationToken cancellationToken)
Avalonia.Threading
Dispatcher.MainLoop (CancellationToken cancellationToken)
Avalonia.Controls.ApplicationLifetimes
ClassicDesktopStyleApplicationLifetime.Start (String[] args)
```

```
System.NullReferenceException: Object reference not set to an instance of an object.

Avalonia.Media.TextFormatting
TextFormatterImpl.MeasureLength (IReadOnlyList`1 textRuns, TextRange textRange, Double paragraphWidth)
Avalonia.Media.TextFormatting
TextFormatterImpl.PerformTextWrapping (List`1 textRuns, TextRange textRange, Double paragraphWidth, TextParagraphProperties paragraphProperties, FlowDirection flowDirection, TextLineBreak currentLineBreak)
Avalonia.Media.TextFormatting
TextLayout.CreateTextLines ()
Avalonia.Controls
TextBlock.CreateTextLayout (Size constraint, String text)
Avalonia.Controls
TextBlock.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
LayoutManager.Measure (ILayoutable control)
Avalonia.Layout
LayoutManager.ExecuteLayoutPass ()
Avalonia.Threading
JobRunner.RunJobs (Nullable`1 priority)
Avalonia.Win32
Win32Platform.WndProc (IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
Avalonia.Win32.Interop
UnmanagedMethods.DispatchMessage (MSG& lpmsg)
Avalonia.Win32
Win32Platform.RunLoop (CancellationToken cancellationToken)
Avalonia.Threading
Dispatcher.MainLoop (CancellationToken cancellationToken)
Avalonia.Controls.ApplicationLifetimes
ClassicDesktopStyleApplicationLifetime.Start (String[] args)
```

```
System.ArgumentOutOfRangeException: Index and length must refer to a location within the string. (Parameter 'length')

System
String.Substring (Int32 startIndex, Int32 length)
Avalonia.Controls
TextBox.HandleTextInput (String input)
Avalonia.Controls
TextBox.OnTextInput (TextInputEventArgs e)
Avalonia.Input.InputElement.<>c
<.cctor>b__32_4 (InputElement x, TextInputEventArgs e)
System.Reactive.Subjects
Subject`1.OnNext (T value) /_/Rx.NET/Source/src/System.Reactive/Subjects/Subject.cs:145
Avalonia.Interactivity
EventRoute.RaiseEventImpl (RoutedEventArgs e)
Avalonia.Interactivity
EventRoute.RaiseEvent (IInteractive source, RoutedEventArgs e)
Avalonia.Interactivity
Interactive.RaiseEvent (RoutedEventArgs e)
Avalonia.Input
KeyboardDevice.ProcessRawEvent (RawInputEventArgs e)
Avalonia.Input
InputManager.ProcessInput (RawInputEventArgs e)
Avalonia.Win32
WindowImpl.AppWndProc (IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
Avalonia.Win32
WindowImpl.WndProc (IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
Avalonia.Win32.Interop
UnmanagedMethods.DispatchMessage (MSG& lpmsg)
Avalonia.Win32
Win32Platform.RunLoop (CancellationToken cancellationToken)
Avalonia.Threading
Dispatcher.MainLoop (CancellationToken cancellationToken)
Avalonia.Controls.ApplicationLifetimes
ClassicDesktopStyleApplicationLifetime.Start (String[] args)
```

```
System.ArgumentNullException: Value cannot be null. (Parameter 'name'

Avalonia.Media.FontFamily
Avalonia.Media.FontFamily..ctor(Uri baseUri, String name)
Avalonia.Media.FontFamily
Avalonia.Media.FontFamily..ctor(String name)
Avalonia.Media.Typeface
Avalonia.Media.Typeface..ctor(String fontFamilyName, FontStyle style, FontWeight weight)
Avalonia.Skia
FontManagerImpl.TryMatchCharacter (Int32 codepoint, FontStyle fontStyle, FontWeight fontWeight, FontFamily fontFamily, CultureInfo culture, Typeface& fontKey)
Avalonia.Media
FontManager.TryMatchCharacter (Int32 codepoint, FontStyle fontStyle, FontWeight fontWeight, FontFamily fontFamily, CultureInfo culture, Typeface& typeface)
Avalonia.Media.TextFormatting
TextCharacters.CreateShapeableRun (ReadOnlySlice`1 text, TextRunProperties defaultProperties, SByte biDiLevel, TextRunProperties& previousProperties)
Avalonia.Media.TextFormatting
TextFormatterImpl.CoalesceLevels (IReadOnlyList`1 textCharacters, ReadOnlySlice`1 levels) +MoveNext()
Avalonia.Media.TextFormatting
TextFormatterImpl.ShapeTextRuns (List`1 textRuns, FlowDirection flowDirection, FlowDirection& resolvedFlowDirection)
Avalonia.Media.TextFormatting
TextLayout.CreateTextLines ()
Avalonia.Controls
TextBlock.CreateTextLayout (Size constraint, String text)
Avalonia.Controls
TextBlock.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Controls
DockPanel.MeasureOverride (Size constraint)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
LayoutHelper.MeasureChild (ILayoutable control, Size availableSize, Thickness padding)
Avalonia.Controls.Presenters
ContentPresenter.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
Layoutable.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
Layoutable.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
LayoutHelper.MeasureChild (ILayoutable control, Size availableSize, Thickness padding)
Avalonia.Controls.Presenters
ContentPresenter.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
Layoutable.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
LayoutHelper.MeasureChild (ILayoutable control, Size availableSize, Thickness padding)
Avalonia.Controls
Border.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
Layoutable.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
LayoutHelper.MeasureChild (ILayoutable control, Size availableSize, Thickness padding)
Avalonia.Controls.Presenters
ContentPresenter.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Controls
StackPanel.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Controls.Presenters
ItemVirtualizer.MeasureOverride (Size availableSize)
Avalonia.Controls.Presenters
ItemsPresenter.MeasureOverride (Size availableSize)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Controls
DockPanel.MeasureOverride (Size constraint)
Avalonia.Layout
Layoutable.MeasureCore (Size availableSize)
Avalonia.Layout
Layoutable.Measure (Size availableSize)
Avalonia.Layout
LayoutManager.Measure (ILayoutable control)
Avalonia.Layout
LayoutManager.Measure (ILayoutable control)
Avalonia.Layout
LayoutManager.Measure (ILayoutable control)
Avalonia.Layout
LayoutManager.Measure (ILayoutable control)
Avalonia.Layout
LayoutManager.Measure (ILayoutable control)
Avalonia.Layout
LayoutManager.ExecuteMeasurePass ()
Avalonia.Layout
LayoutManager.InnerLayoutPass ()
Avalonia.Layout
LayoutManager.ExecuteLayoutPass ()
Avalonia.Threading
JobRunner.RunJobs (Nullable`1 priority)
Avalonia.Win32
Win32Platform.WndProc (IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
Avalonia.Win32.Interop
UnmanagedMethods.DispatchMessage (MSG& lpmsg)
Avalonia.Win32
Win32Platform.RunLoop (CancellationToken cancellationToken)
Avalonia.Threading
Dispatcher.MainLoop (CancellationToken cancellationToken)
Avalonia.Controls.ApplicationLifetimes
ClassicDesktopStyleApplicationLifetime.Start (String[] args)
```

**Comment by Gillibald** [comment]
@adirh3 Is there dynamic text displayed by TextBlocks in your app or is it just static? Ideally those users could give me more info about the actual text they were typing or what they were doing at the moment of the crash.

**Comment by adirh3** [comment]
> @adirh3 Is there dynamic text displayed by TextBlocks in your app or is it just static? Ideally those users could give me more info about the actual text they were typing or what they were doing at the moment of the crash.

I believe most of these are from using TextBox's, and 99% of the displayed text is produced from bindings (TextBlocks).
I will try asking them again to post reproes if they find a way.

**Comment by hez2010** [comment]
Apparently last changes broke something with alignment.

![image](https://user-images.githubusercontent.com/14960345/155469729-17c2ba88-a1ab-4f9c-836c-f9bb75f47c38.png)

**Comment by Gillibald** [comment]
![Screenshot 2022-02-24 074244](https://user-images.githubusercontent.com/5928797/155472112-7a25cc54-1438-456d-b66d-e4283d3a4174.png)

Build with a just clones master

**Comment by hez2010** [comment]
> Are you using the nuget feed?

I pulled code from master branch and build it directly.

**Comment by Gillibald** [comment]
I just did this:
```
git clone https://github.com/AvaloniaUI/Avalonia.git
git submodule init
git submodule update
cd Avalonia\samples\ControlCatalog.NetCore
dotnet run
```
And I get the above result. I am also on Windows 11.

**Comment by hez2010** [comment]
I found how to repro it: open ControlCatalog, then change your screen dpi.

**Comment by danipen** [comment]
Yes, reproduce it here too. It seems that DPI resolution affects to the text vertical alignment:
- 100% DPI:
![image](https://user-images.githubusercontent.com/501613/155477853-3b0eed6a-1f88-4c75-81f7-906863519c93.png)
- 125% DPI:
![image](https://user-images.githubusercontent.com/501613/155477896-2e829a9b-bf98-4252-b1cc-94cdbc328ae2.png)
- 350% DPI:
![image](https://user-images.githubusercontent.com/501613/155477945-9164d8bf-0ac3-4255-9920-aaf5e9152c4a.png)

I found a couple more:
- Hitting <kbd>End</kbd> the caret is not visible:
![textbox-end](https://user-images.githubusercontent.com/501613/155478215-0fb54d12-9aa9-41bb-866f-d37a1bf37216.gif)
- Word navigation/selection doesn't work: <kbd>Ctrl + left</kbd>, <kbd>Ctrl + right</kbd>, <kbd>Ctrl + Shift + left</kbd>, <kbd>Ctrl + Shift + right</kbd>

**Comment by Gillibald** [comment]
Ahh okay so I guess unless I scale up the text it doesn't make sense to scale the bounding box

**Comment by Gillibald** [comment]
#7693 should fix scaling issues

**Comment by adirh3** [comment]
after #7693 -

Some selection bugs - 
1. Type something
2. Press Ctrl+Shift+Left
3. See selection not full
![image](https://user-images.githubusercontent.com/27368554/155569369-b7d4d080-160a-4b7e-9d93-59840b6bf7d9.png)
4. Press backspace, get this exception -
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'length')
   at Avalonia.Utilities.ReadOnlySlice`1.Take(Int32 length)
   at Avalonia.Media.TextFormatting.FormattedTextSource.CreateTextStyleRun(ReadOnlySlice`1 text, TextRunProperties defaultProperties, IReadOnlyList`1 textModifier)
   at Avalonia.Media.TextFormatting.FormattedTextSource.GetTextRun(Int32 textSourceIndex)
   at Avalonia.Media.TextFormatting.TextFormatterImpl.TextRunEnumerator.MoveNext()
   at Avalonia.Media.TextFormatting.TextFormatterImpl.FetchTextRuns(ITextSource textSource, Int32 firstTextSourceIndex, TextEndOfLine& endOfLine, TextRange& textRange)
   at Avalonia.Media.TextFormatting.TextFormatterImpl.FormatLine(ITextSource textSource, Int32 firstTextSourceIndex, Double paragraphWidth, TextParagraphProperties paragraphProperties, TextLineBreak previousLineBreak)
   at Avalonia.Media.TextFormatting.TextLayout.CreateTextLines()
   at Avalonia.Media.TextFormatting.TextLayout..ctor(String text, Typeface typeface, Double fontSize, IBrush foreground, TextAlignment textAlignment, TextWrapping textWrapping, TextTrimming textTrimming, TextDecorationCollection textDecorations, FlowDirection flowDirection, Double maxWidth, Double maxHeight, Double lineHeight, Int32 maxLines, IReadOnlyList`1 textStyleOverrides)
   at Avalonia.Controls.Presenters.TextPresenter.CreateTextLayoutInternal(Size constraint, String text, Typeface typeface, IReadOnlyList`1 textStyleOverrides)
   at Avalonia.Controls.Presenters.TextPresenter.CreateTextLayout()
   at Avalonia.Controls.Presenters.TextPresenter.get_TextLayout()
   at Avalonia.Controls.Presenters.TextPresenter.MoveCaretToTextPosition(Int32 textPosition, Boolean trailingEdge)
   at Avalonia.Controls.Presenters.TextPresenter.set_CaretIndex(Int32 value)
   at Avalonia.Controls.Presenters.TextPresenter.<>c.<.cctor>b__23_1(TextPresenter o, Int32 v)
   at Avalonia.DirectProperty`2.InvokeSetter(IAvaloniaObject instance, BindingValue`1 value)
   at Avalonia.AvaloniaObject.SetDirectValueUnchecked[T](DirectPropertyBase`1 property, BindingValue`1 value)
   at Avalonia.AvaloniaObject.DirectBindingSubscription`1.OnNext(BindingValue`1 value)
   at Avalonia.Reactive.SingleSubscriberObservableBase`1.PublishNext(T value)
   at Avalonia.Reactive.TypedBindingAdapter`1.OnNext(BindingValue`1 value)
   at Avalonia.Reactive.SingleSubscriberObservableBase`1.PublishNext(T value)
   at Avalonia.Reactive.BindingValueAdapter`1.OnNext(T value)
   at Avalonia.Reactive.SingleSubscriberObservableBase`1.PublishNext(T value)
   at Avalonia.Data.TemplateBinding.PublishValue()
   at Avalonia.Data.TemplateBinding.TemplatedParentPropertyChanged(Object sender, AvaloniaPropertyChangedEventArgs e)
   at Avalonia.AvaloniaObject.RaisePropertyChanged[T](AvaloniaPropertyChangedEventArgs`1 change)
   at Avalonia.AvaloniaObject.RaisePropertyChanged[T](AvaloniaProperty`1 property, Optional`1 oldValue, BindingValue`1 newValue, BindingPriority priority)
   at Avalonia.AvaloniaObject.SetAndRaise[T](AvaloniaProperty`1 property, T& field, T value)
   at Avalonia.Controls.TextBox.set_CaretIndex(Int32 value)
   at Avalonia.Controls.TextBox.DeleteSelection(Boolean raiseTextChanged)
   at Avalonia.Controls.TextBox.OnKeyDown(KeyEventArgs e)
   at Blast.Controls.CustomTextBox.OnKeyDown(KeyEventArgs e) in C:\Git\Blast\Blast\Controls\CustomTextBox.cs:line 14
   at Avalonia.Input.InputElement.<>c.<.cctor>b__32_2(InputElement x, KeyEventArgs e)
   at Avalonia.Interactivity.RoutedEvent`1.<>c__DisplayClass2_0`1.<AddClassHandler>g__Adapter|0(Object sender, RoutedEventArgs e)
   at Avalonia.Interactivity.RoutedEvent.<>c__DisplayClass23_0.<AddClassHandler>b__0(ValueTuple`2 args)
   at System.Reactive.AnonymousObserver`1.OnNextCore(T value) in /_/Rx.NET/Source/src/System.Reactive/AnonymousObserver.cs:line 67
   at System.Reactive.ObserverBase`1.OnNext(T value) in /_/Rx.NET/Source/src/System.Reactive/ObserverBase.cs:line 34
   at System.Reactive.Subjects.Subject`1.OnNext(T value) in /_/Rx.NET/Source/src/System.Reactive/Subjects/Subject.cs:line 147
   at Avalonia.Interactivity.RoutedEvent.InvokeRaised(Object sender, RoutedEventArgs e)
   at Avalonia.Interactivity.EventRoute.RaiseEventImpl(RoutedEventArgs e)
   at Avalonia.Interactivity.EventRoute.RaiseEvent(IInteractive source, RoutedEventArgs e)
   at Avalonia.Interactivity.Interactive.RaiseEvent(RoutedEventArgs e)
   at Avalonia.Input.KeyboardDevice.ProcessRawEvent(RawInputEventArgs e)
   at Avalonia.Input.InputManager.ProcessInput(RawInputEventArgs e)
   at Avalonia.Controls.TopLevel.HandleInput(RawInputEventArgs e)
   at Avalonia.Win32.WindowImpl.AppWndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
   at Avalonia.Win32.WindowImpl.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
   at Avalonia.Win32.Interop.UnmanagedMethods.DispatchMessage(MSG& lpmsg)
   at Avalonia.Win32.Win32Platform.RunLoop(CancellationToken cancellationToken)
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
   at Blast.Program.Main(String[] args) in C:\Git\Blast\Blast\Program.cs:line 263

**Comment by danipen** [comment]
@Gillibald I'm testing the recently merged #7693, and I can still see a couple of issues:
- Hitting <kbd>Shift + End</kbd> in the first line works fine. Doing that in the second line doesn't.
![tb-issue1](https://user-images.githubusercontent.com/501613/156619578-6da4a64f-e9cf-4eb1-916d-b2f7e82f7ef6.gif)

- Using <kbd>Shift + Ctrl + Right</kbd> doesn't move the horizontal scroll right when the caret gets out of the viewport (however using standard cursor movement with keys works fine).
![tb-issue2](https://user-images.githubusercontent.com/501613/156619664-bfe33f17-3794-46be-93a2-7fc32665075e.gif)

**Comment by Gillibald** [comment]
Thanks for you feedback will have a look

**Comment by Gillibald** [comment]
https://github.com/AvaloniaUI/Avalonia/pull/7746 should fix your last findings

**Comment by danipen** [comment]
Yes, now issues are fixed, thanks!!

I found a couple of new more, but really small ones:
- <kbd>Shift + Ctrl + End</kbd> + <kbd>Home</kbd>: Caret jumps one line up.
![b-1](https://user-images.githubusercontent.com/501613/156728860-59eba4a5-d14e-4ece-bbeb-2a221dbd239b.gif)
- <kbd>Shift + Ctrl + End</kbd> + <kbd>Shift + Home</kbd>: Caret disappears.

**Comment by hez2010** [comment]
I found another bug:
1. input an emoji using win+.
2. close the emoji select panel
3. shift + left

![image](https://user-images.githubusercontent.com/14960345/156866763-56486489-f24e-4e14-9001-62ab7666920f.png)

**Comment by Gillibald** [comment]
It seems to split the surrogate pair

**Comment by danielmayost** [comment]
if `FlowDirection="RightToLeft` on `TextBox` then `Backwards` key isn't responding or responding wrong.

**Comment by Gillibald** [comment]
> if `FlowDirection="RightToLeft` on `TextBox` then `Backwards` key isn't responding or responding wrong.

Wil have a look

**Comment by robloo** [comment]
In control catalog's Buttons page TextWrapping seems to somehow be broken only for certain text. This was working a month or so ago so I assume it's the TextFormatting changes. Strangely, the exact same XAML is working just fine for slightly different text.

![image](https://user-images.githubusercontent.com/17993847/159134888-c41d32d4-df85-4588-8a85-9e7253e83fe8.png)

**Comment by danielmayost** [comment]
Some issues I encountered during my work #7810:
- [ ] as @robloo said, many broken text in `ControlCatalog`, even within `ToolTip`, I guess this is the same problem.
- [ ] text in `Flyouts` Tab exits outside the right border if is RTL FlowDirection:
<img src="https://user-images.githubusercontent.com/41772276/159263593-f20d4efd-7f50-40e0-9b6d-7bb5b52bf947.png" height=180 width=220/>
- [ ] If you change the `FlowDirection` to `TextBox` while running, the text is not aligned to the appropriate side, only if you click on the text it aligns the text:
<img src="https://user-images.githubusercontent.com/41772276/159265115-7227d71a-e5f7-46fe-ab88-64e54212064c.gif" height=30width=150/>

**Comment by danipen** [comment]
@Gillibald, I found a missing feature. 

For all textboxes implementation I checked, **double-clicking and dragging over text, allows you to extend the selection by word.**

### Example
https://user-images.githubusercontent.com/501613/159515582-c0a4f7e0-6e88-4799-a469-2a93a121fb4d.mov

### Avalonia's
https://user-images.githubusercontent.com/501613/159515549-88039fd5-3085-4427-8d2f-f200fa570aad.mov

**Comment by danipen** [comment]
- [ ] Hitting <kbd>End</kbd> in the following scenario places caret in the previous paragraph.
![tb-issue](https://user-images.githubusercontent.com/501613/159516532-2d31bd10-f9b1-4fd8-b392-39340cb1a47f.gif)

**Comment by robloo** [comment]
@Gillibald A lot of issues are related to the `TextBox` functionality I notice. Would it be a good idea to just port the [WPF TextBox](https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/TextBox.cs) to Avalonia now? I wouldn't think this would be too difficult considering all the lower-level text API's are now implemented. I know that wasn't possible before but now that WPF is open source and all the base pieces are there this makes the most sense to me.

**Comment by danielmayost** [comment]
@Gillibald I saw that you marked that the problems were fixed, I checked in front of your industry and did not see any change, you have not yet pushed the changes?

**Comment by Gillibald** [comment]
@robloo Porting WPF's TextBox isn't possible at the moment because it requires the TextContainer. For reference: https://github.com/AvaloniaUI/Avalonia/blob/inlinesTheThird/src/Avalonia.Documents/Documents/TextContainer.cs

**Comment by danipen** [comment]
> I can't reproduce your last issue. Might be fixed already.

@Gillibald, sorry I still reproduce the issue in **LAST** changes on master (commit `[0483f9](https://github.com/AvaloniaUI/Avalonia/commit/0483f9bced0ea444b4b25e8b69659e54b80157e8)`). **Should I test different changes?**

The scenario is just clicking the second paragraph and hitting <kbd>End/<kbd> key:

https://user-images.githubusercontent.com/501613/160274457-f1ec9f2e-222b-4288-8c89-5ae5da29b869.mp4

**Comment by danielmayost** [comment]
- [ ] when RTL FlowDirection, text within tooltip get wrong:
![image](https://user-images.githubusercontent.com/41772276/160542166-79354d61-1c43-45ed-8dd2-3a5f08f723d7.png)

**Comment by Gillibald** [comment]
@danipen You always need to test against: https://github.com/AvaloniaUI/Avalonia/pull/7746

**Comment by robloo** [comment]
@Gillibald 

[ ] I'm seeing multiple instances where text length is not being measured correctly with latest master. The render length does not match the measured length somehow. This may be related to only certain characters like 'W'.

 * The tooltip is set to "Magenta"
 ![image](https://user-images.githubusercontent.com/17993847/165657594-020aacbc-1799-4129-884e-a442c116a065.png)
 * The ComboBox text content is so long it's cutting off the right border
 ![image](https://user-images.githubusercontent.com/17993847/165657733-72692b42-9fb2-4b2c-b177-0fc960d164b5.png)
 * CalendarDatePicker text with a 'W' (Wednesday) seems to consistently cause this. Other dates without a 'W' are measured correctly.
 ![image](https://user-images.githubusercontent.com/17993847/165657923-8365c4d1-27b3-415c-9436-f5de9bfad292.png)

**Comment by Gillibald** [comment]
Please test against: https://github.com/AvaloniaUI/Avalonia/pull/7946

0.10.999-cibuild0020141-beta

**Comment by danielmayost** [comment]
@Gillibald there is some new bugs since #7946 merged.
just enter to `TextBox` tab in ControlCatalog and try to select some text.

**Comment by danielmayost** [comment]
- [ ] [RTL] when `TextBox` is empty, the cursor is displayed incorrectly on the left side.
- [ ] [RTL] in `Multiline TextBox`, typing enter after `א` to new line  is not respected (the cursor not move).

**Comment by Gillibald** [comment]
I will add some unit tests that cover measurements with scaling

**Comment by robloo** [comment]
@Gillibald @danielmayost  That's a good point, I should add I'm running Windows with 125% scaling. That may account for some things above but I will retest also (I'm sure @danielmayost is correct though).

I noticed Avalonia doesn't handle bitmaps the same as WPF/WinUI and will always render them to the physical device pixel DPI. That required introducing two coordinate systems in the ColorPicker which isn't required in UWP. It's possible the rasterization step of text here is introducing similar inconsistencies. There is likely a deeper discussion required for how Avalonia should handle these things. Fundamentally, it seems WPF/UWP scale bitmaps automatically to match the app coordinate system and any desktop scaling factor.

**Comment by Gillibald** [comment]
This is with 125% scaling
![Screenshot 2022-04-28 205708](https://user-images.githubusercontent.com/5928797/165826344-efc72098-88c1-47b8-8b54-bc8dd746d056.png)

**Comment by robloo** [comment]
> This is with 125% scaling

Does this mean you fixed the issue or can't reproduce it? I noticed it is only certain text strings. Switch to an EN culture and try Wednesday dates.

**Comment by danielmayost** [comment]
175%:
![image](https://user-images.githubusercontent.com/41772276/165829026-a6a735df-338f-4422-a91c-71421272701c.png)
225%:
![image](https://user-images.githubusercontent.com/41772276/165829097-fc600378-21b8-4353-8d4f-c3c4fd8edbb8.png)

**Comment by danielmayost** [comment]
Strange, maybe it's a combination of scale and resolution, mine 3840 * 2160 and 175% scaling, in ControlCatalog's ToolTip ControlCatalog It happens.

**Comment by robloo** [comment]
I tested again on latest master and confirmed all issues are still present even after the InlineUIContainer changes. I'm not sure how to repro this if it doesn't show up on your computer though. Very strange.

**Comment by Gillibald** [comment]
I can repro the issue with 175% scaling!

**Comment by adirh3** [comment]
@Gillibald I tried playing around with it with different DPIs/Resolutions and found some issues -

Using this tooltip instead of the `This is a tooltip` in the control catalog -
```
     <ToolTip.Tip>
                <TextBlock FontFamily="Segoe UI Variable Text Light" FontSize="15.120000000000001">Nahimic</TextBlock>
      </ToolTip.Tip>
```
With monitor resolution 3840x2160, 150% scaling -
<img width="147" alt="image" src="https://user-images.githubusercontent.com/27368554/165960706-272bfc90-257f-4684-b296-258d44b3d0ff.png">

After changing line 629 in `TextBlock` to this -
`var textSize = PixelSize.FromSize(finalSize.Deflate(padding), VisualRoot?.RenderScaling ?? 1);`

<img width="154" alt="image" src="https://user-images.githubusercontent.com/27368554/165960522-cc41b1e9-bc4f-4292-b90a-3b523e656a75.png">

Other issue with empty line in tooltip (same monitor) - 
![image](https://user-images.githubusercontent.com/27368554/165960784-ad5a4851-bb56-4546-ac58-8725cfe65da7.png)

Can't repro anything outside tooltips.
On 100% scaling there aren't any issues from my experience, so maybe need to add the RenderScaling to other places as well?

**Comment by Gillibald** [comment]
I have tested all scalings against your proposed change and it looks like there are no issues anymore.

**Comment by robloo** [comment]
I also confirmed @adirh3 's fix worked for TextBlock and tooltips. TextBox however still has issues and I'm still seeing the issue in CalendarDatePicker for Wednesday dates. ComboBox also still has a problem but it may be unrelated.

**Comment by adirh3** [comment]
@Gillibald, there are still some issues when setting `TextWrapping="Wrap"` -

```
<TextBlock FontFamily="Segoe UI Variable Text Light"
                           TextWrapping="Wrap"
                           FontSize="15.120000000000001" Text="Introduction to Microsoft Edge WebView2 - Microsoft Edge Development | Microsoft Docs"/>
```

![image](https://user-images.githubusercontent.com/27368554/165981437-014d6e65-88f4-4f17-afa3-d3d625667236.png)
![image](https://user-images.githubusercontent.com/27368554/165981562-0d2881ba-4bf9-46cf-ab1f-cde0935f94c3.png)

**Comment by danielmayost** [comment]
> I can't repro your CalendarDatePicker issue.

@Gillibald try to display `Wednesday, 25 May 2022`, this issue not related to scaling:
![image](https://user-images.githubusercontent.com/41772276/166119844-720c5c99-1cce-4200-95a5-c9c730e03d08.png)

**Comment by Gillibald** [comment]
The CalendarDatePicker issue isn't related to text rendering. THe issue is that the maximum with of the parent StackPanel is 200 and therefore the text does not fit anymore.
![Screenshot 2022-05-02 113245](https://user-images.githubusercontent.com/5928797/166215617-cfaa3b70-4d8d-4b60-954e-305df49c13c1.png)

**Comment by robloo** [comment]
> The issue is that the maximum width of the parent StackPanel is 200

Wow, good find. I guess I'm over cautious with text processing issues! I'll review WinUI/WPF and then add a fix to another PR. Either we should have ellipsis or allow all width.

**Comment by Gillibald** [comment]
#8070
